Sources/_Datadog_Private/Swizzler.m

```diff
@@ -0,0 +1,212 @@
+/*
+* Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+* This product includes software developed at Datadog (https://www.datadoghq.com/).
+* Copyright 2019-2020 Datadog, Inc.
+*/
+
+#import "Swizzler.h"
+@import ObjectiveC.runtime;
+@import ObjectiveC.message;
+
+/*
+ TemplateURLSession:
+    • overridden methods: dataTaskWithURL:, dataTaskWithURLRequest:, etc.
+    • compile-time placeholder methods: injected_interceptRequest:, injected_observeTask:, super_dataTaskWithURLRequest:, etc.
+        1. injected_ methods
+        2. super_ methods
+ Swizzler:
+    1. take object
+    2. lookup dynamic class; early return if exists
+    3. create if not exists
+        • add instance methods from TemplateURLSession
+        • add injected methods from blocks passed as parameters
+    4. change class of object
+
+ Open questions:
+ 1. what to do with URLTaskPublisher? it's Swift struct, un-swizzable
+ 2. profile dynamic class, does it add overhead beyond limits?
+ 3. thread safety in swizzling?
+ */
+
+// IMPORTANT NOTE: TemplateURLSession should NOT have "ivar"s other than what NSURLSession has
+// Ivars change the memory layout of the class and needs to be added very carefully unless if it MUST be added
+@interface TemplateURLSession : NSURLSession
+@end
+
+// Placeholder methods are NEVER used at runtime
+// Their implementations at runtime either come from superclass or injected by Swizzler
+@interface TemplateURLSession (PlaceholderMethods)
+- (NSURLRequest *)injected_interceptRequest:(NSURLRequest *)request;
+- (void)injected_observeTask:(NSURLSessionTask *)task;
+- (NSURLSessionDataTask *)super_dataTaskWithRequest:(NSURLRequest *)request;
+@end
+
+@implementation TemplateURLSession
+- (NSURLSessionDataTask *)dataTaskWithURL:(NSURL *)url {
+    NSURLRequest *request = [NSURLRequest
+                             requestWithURL:url
+                             cachePolicy:self.configuration.requestCachePolicy
+                             timeoutInterval:self.configuration.timeoutIntervalForRequest];
+    NSURLRequest *modifiedRequest = [self injected_interceptRequest:request];
+    /*
+     IMPORTANT NOTE: "super" keyword is resolved at compile-time,
+     so [super dataTaskWithRequest:modifiedRequest] would resolve to NSURLSession.dataTaskWithRequest:
+     as NSURLSession is the superclass of this class at compile-time.
+     However, superclass of this method may be different at runtime.
+     Eg: [NSURLSession shared] is _NSURLLocalSession, which is different than NSURLSession.
+     */
+    NSURLSessionDataTask *dataTask = [self super_dataTaskWithRequest:modifiedRequest];
+    [self injected_observeTask:dataTask];
```

Comment:
`DDSpan` will be created in `injected_observeTask`, but the headers must be modified in `injected_interceptRequest` which is called before. I think this will not let us pass the `trace_id`, no?

Reply:
I was thinking about this for a while. If it's not possible to follow tracer's public API order:
```swift
// on request started
let span = tracer.startSpan()

// on request mutation
let traceHeaders = DDHTTPHeadersWriter()
Global.sharedTracer.inject(spanContext: span.context, writer: traceHeaders)
request.allHTTPHeaderFields += traceHeaders.tracePropagationHTTPHeaders

// on request completed
span.finish()
```

we might add some internal APIs to follow the `URLSession` order:
```swift
// on request mutation
let spanContext = DDSpanContext(...)
let traceHeaders = DDHTTPHeadersWriter()
Global.sharedTracer.inject(spanContext: spanContext, writer: traceHeaders)
request.allHTTPHeaderFields += traceHeaders.tracePropagationHTTPHeaders

// on request started
let span = tracer.startSpan(with: spanContext) // internal tracer API

// on request completed
span.finish()
```

However, if possible, that would be nice to have only one tracer API to use 🙂.

Reply:
This is a major issue that we need to discuss, I create another ticket for this